src/proxy_protocol/mod.rs

```diff
@@ -0,0 +1,571 @@
+            "No valid V1 Proxy Protocol header received",
+        )),
+        HeaderResult::V2(Err(_error)) => Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "No valid V2 Proxy Protocol header received",
+        )),
+    }
+}
+
+/// Middleware for adding client IP address to the request `forwarded` header.
+/// see spec: <https://www.rfc-editor.org/rfc/rfc7239#section-5.2>
+#[derive(Debug)]
+pub struct ForwardClientIp<S> {
+    inner: S,
+    client_address_opt: Option<SocketAddr>,
+}
+
+impl<S> Service<Request<hyper::Body>> for ForwardClientIp<S>
+where
+    S: Service<Request<hyper::Body>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<hyper::Body>) -> Self::Future {
+        let forwarded_string = match self.client_address_opt {
+            Some(socket_addr) => match socket_addr {
+                SocketAddr::V4(addr) => {
+                    format!("for={}:{}", addr.ip(), addr.port())
+                }
+                SocketAddr::V6(addr) => {
+                    format!("for=\"[{}]:{}\"", addr.ip(), addr.port())
+                }
+            },
+            None => "for=unknown".to_string(),
+        };
+
+        if let Ok(header_value) = HeaderValue::from_str(&forwarded_string) {
+            req.headers_mut().insert("Forwarded", header_value);
+        }
+
+        self.inner.call(req)
+    }
+}
+
+/// Acceptor wrapper for receiving Proxy Protocol headers.
+#[derive(Clone)]
+pub struct ProxyProtocolAcceptor<A> {
+    inner: A,
+    parsing_timeout: Duration,
+}
+
+impl<A> ProxyProtocolAcceptor<A> {
+    /// Create a new proxy protocol acceptor from an initial acceptor.
+    /// This is compatible with tls acceptors.
+    pub fn new(inner: A) -> Self {
+        #[cfg(not(test))]
+        let parsing_timeout = Duration::from_secs(10);
+
+        // Don't force tests to wait too long.
+        #[cfg(test)]
+        let parsing_timeout = Duration::from_secs(1);
+
+        Self {
+            inner,
+            parsing_timeout,
+        }
+    }
+
+    /// Override the default Proxy Header parsing timeout of 10 seconds, except during testing.
+    pub fn parsing_timeout(mut self, val: Duration) -> Self {
+        self.parsing_timeout = val;
+        self
+    }
+}
+
+impl<A> ProxyProtocolAcceptor<A> {
+    /// Overwrite inner acceptor.
+    pub fn acceptor<Acceptor>(self, acceptor: Acceptor) -> ProxyProtocolAcceptor<Acceptor> {
+        ProxyProtocolAcceptor {
+            inner: acceptor,
+            parsing_timeout: self.parsing_timeout,
+        }
+    }
+}
+
+impl<A, I, S> Accept<I, S> for ProxyProtocolAcceptor<A>
+where
+    A: Accept<I, S> + Clone,
+    A::Stream: AsyncRead + AsyncWrite + Unpin,
+    I: AsyncRead + AsyncWrite + Unpin + Send + 'static,
```

Comment:
why static?

Reply:
lifetime of stream

Reply:
```
error[E0310]: the parameter type `I` may not live long enough
   --> src/proxy_protocol/mod.rs:271:43
    |
271 |             timeout(self.parsing_timeout, future),
    |                                           ^^^^^^ ...so that the type `I` will meet its required lifetime bounds
    |
help: consider adding an explicit lifetime bound...
    |
256 |     I: AsyncRead + AsyncWrite + Unpin + Send + 'static,
    |                                              +++++++++


```